src/cpp/core/include/core/r_util/RActiveSessionStorage.hpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * RActiveSessionStorage.hpp
+ *
+ * Copyright (C) 2021 by Rstudio, PBC
+ *
+ * Unless you have received this program directly from Rstudio pursuant
+ * to the terms of a commercial license agreement with Rstudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#ifndef CORE_R_UTIL_R_ACTIVE_SESSION_STORAGE
+#define CORE_R_UTIL_R_ACTIVE_SESSION_STORAGE
+
+#include <shared_core/FilePath.hpp>
+
+namespace rstudio {
+namespace core {
+namespace r_util {
+
+   class IActiveSessionStorage 
+   {
+   public:
+      Error virtual readProperty(const std::string& id, const std::string& name, std::string* pValue) = 0;
+      Error virtual writeProperty(const std::string& id, const std::string& name, const std::string& value) = 0;
+
+   protected:
+      virtual ~IActiveSessionStorage() = default;
+      IActiveSessionStorage() = default;
+   };
+
+   class LegacySessionStorage : public IActiveSessionStorage
+   {
+   public:
+      explicit LegacySessionStorage(const FilePath& location);
+
+      Error readProperty(const std::string& id, const std::string& name, std::string* pValue) override;   
+      Error writeProperty(const std::string& id, const std::string& name, const std::string& value) override;
+
+   private:
+      FilePath activeSessionsDir_;
+      const std::string propertiesDirName_ = "properites";
+      const std::string legacySessionDirPrefix_ = "session-";
+
+      FilePath buildPropertyPath(const std::string& id, const std::string& name);
+
+      static const std::string& getLegacyName(const std::string& name)
+      {
+         static const std::map<std::string, std::string> legacyNames = 
+         {
+            { "last_used" , "last-used" },
+            { "r_version" , "r-version" },
+            { "r_version_label" , "r-version-label" },
+            { "r_version_home" , "r-version-home" },
+            { "working_directory" , "working-dir" },
+            { "launch_parameters" , "launch-parameters" }
+         };
+
+         if (legacyNames.find(name) != legacyNames.end())
+            return legacyNames.at(name);
+
+         return name;
+      }
+   };
+
+   class ActiveSessionStorageFactory
+   {
+   public:
+      static std::shared_ptr<IActiveSessionStorage> getActiveSessionStorage();
```

Comment:
What's the purpose of returning shared pointers here, as opposed to unique pointers? With shared pointers, we'll need to consider multi-threading, but as I think about this more, maybe we need to consider multi-threading anyway, since the file can change on disk while another thread is reading it, even if you're not using shared pointers (i.e if two storages are in use for the same session).

What do you think about a minor refactor so that the session storage accepts an ID, the `IActiveSessionStorage` becomes `AbstractSessionStorage`, it keeps the ID in a protected member, and it uses a ReaderWriterMutex (there's a pre-existing class for this somewhere in core or shared_core), and calls into the private, inheritable read/write functions after locking?

Then the factory would keep a map of session id->shared_ptr (for which you will also need a mutex, but a regular one will do) and if the storage for that session already exists, return it instead of creating a new one. The only thing left to consider is how to clean unused sessions out of the map. The most self contained way would be to use [`std::shared_ptr::unique`](https://en.cppreference.com/w/cpp/memory/shared_ptr/unique) to check whether you're the only one referencing that object and delete it from the map, if so. That could either be done as part of any invocation of `getActiveSessionStorage` or `getLegacyActiveSessionStorage` or on a regular, timed callback.

Reply:
I'll take more time to look through and let LegacySessionStorage to become more complicated. In support of future database and rpc request activities, factory could produce a single instance of SessionStorage. This would necessitate that the session storage object concerns itself with thread safety.

LegacySessionStorage would need a mutex per session id. May not matter, session writes / server reads. Investigate
Think about this more.

Reply:
Apologies for chiming in late here or repeating these comments. I thought I already made them but don't see them here now.

- As far as I can tell, the SessionStorage object itself is ok to be global/static - i.e. we only use it from the session and the fields never change. So I agree with Maria that we don't need the shared_ptr since that implies multiple clients.  
- I don't have a recommendation for fixing this now, but it's unfortunate from the performance perspective that we have readProperty/writeProperty calls, without any batching - i.e. no way to call readProperties and writeProperties.  With the DB version, each of these will turn into round-trips of the pipeline: rsession -> rserver -> postgres -> commit.  Right now, it's bad enough with rsession -> nfs-file and we are going to make it worse. If each hop is 10ms, 3 calls in a row is only 30ms but if latency for any hop gets high (like 200ms) it can really slow things down. Maybe the RPC apis should support multiple properties and we can make it easier for performance tuning?  I suppose it's also something we could address later...

Reply:
@jeffvroom I'm glad you brought that up, I've been meaning to ping you, but so many other things cropped up. One thing that I have hoped to do, and it's influencing the thinking around the FileActiveSessionStorage, is that database or rpc instances should be used by multiple clients, and encapsulate a batching mechanism to reduce the round trip calls, and the number of open connections to the database or the server. I wanted to ask you for your opinions, I'm worried about the latency, especially to the remote rpc, and the open connections to the database when we have multiple sessions involved.

The rpc apis already are specified to read or write multiple properties at a time, so that will reduce the number of round trips necessary, but it's currently only able to batch together changes per session, I don't know if that's enough batching, or if we need to go further.